packages/customize-widgets/src/components/sidebar-block-editor/use-sidebar-block-editor.js

```diff
@@ -160,107 +158,105 @@ export default function useSidebarBlockEditor( sidebar ) {
 
-			const blocksByWidgetId = keyBy(
-				state.blocks,
-				( block ) => block.attributes.__internalWidgetId
-			);
+			setState( ( lastState ) => {
+				const blocksByWidgetId = keyBy(
+					lastState,
+					( block ) => block.attributes.__internalWidgetId
+				);
 
-			const nextBlocks = _nextBlocks.map( ( nextBlock, index ) => {
-				if (
-					nextBlock.attributes.__internalWidgetId &&
-					nextBlock.attributes.__internalWidgetId in blocksByWidgetId
-				) {
-					const block =
-						blocksByWidgetId[
-							nextBlock.attributes.__internalWidgetId
-						];
-					if ( ! isEqual( block, nextBlock ) ) {
-						const widgetId =
-							nextBlock.attributes.__internalWidgetId;
-						const widgetToUpdate = sidebar.getWidget( widgetId );
-						const widget = blockToWidget(
-							nextBlock,
-							widgetToUpdate
-						);
-						sidebar.updateWidget( widget );
+				const nextBlocks = _nextBlocks.map( ( nextBlock, index ) => {
+					if (
+						nextBlock.attributes.__internalWidgetId &&
+						nextBlock.attributes.__internalWidgetId in
+							blocksByWidgetId
+					) {
+						const block =
+							blocksByWidgetId[
+								nextBlock.attributes.__internalWidgetId
+							];
+						if ( ! isEqual( block, nextBlock ) ) {
+							const widgetId =
+								nextBlock.attributes.__internalWidgetId;
+							const widgetToUpdate = sidebar.getWidget(
+								widgetId
+							);
+							const widget = blockToWidget(
+								nextBlock,
+								widgetToUpdate
+							);
+							sidebar.updateWidget( widget );
+						}
+						return nextBlock;
 					}
-					return nextBlock;
-				}
-
-				const widget = blockToWidget( nextBlock );
-				const widgetId = sidebar.addWidget( widget, index );
-				return {
-					...nextBlock,
-					attributes: {
-						...nextBlock.attributes,
-						__internalWidgetId: widgetId,
-					},
-				};
-			} );
 
-			const seen = nextBlocks.map(
-				( block ) => block.attributes.__internalWidgetId
-			);
+					const widget = blockToWidget( nextBlock );
+					const widgetId = sidebar.addWidget( widget, index );
+					return {
+						...nextBlock,
+						attributes: {
+							...nextBlock.attributes,
+							__internalWidgetId: widgetId,
+						},
+					};
+				} );
+
+				const seen = nextBlocks.map(
+					( block ) => block.attributes.__internalWidgetId
+				);
 
-			for ( const block of state.blocks ) {
-				const widgetId = block.attributes.__internalWidgetId;
-				if ( ! seen.includes( widgetId ) ) {
-					sidebar.removeWidget( widgetId );
+				for ( const block of lastState ) {
+					const widgetId = block.attributes.__internalWidgetId;
+					if ( ! seen.includes( widgetId ) ) {
+						sidebar.removeWidget( widgetId );
+					}
 				}
-			}
 
-			if (
-				nextBlocks.length === state.blocks.length &&
-				! isEqual(
-					nextBlocks.map(
-						( { attributes: { __internalWidgetId } } ) =>
-							__internalWidgetId
-					),
-					state.blocks.map(
+				if (
+					nextBlocks.length === lastState.length &&
+					! isEqual(
```

Comment:
Yeah, true! We'll probably refactor this sometimes soon though. Since it's not changed in this PR then let's just leave it for now, unless there's a bug in it 😉.